Add (random) seed api

WaffleLapkin opened the pull request:
tl;dr: this PR adds the following APIs:
```rust
impl FxHasher {
    pub fn with_seed(seed: usize) -> FxHasher;
}

// Everything below is under cfg(feature = "rand")

pub struct FxRandomState { ... }

impl FxRandomState {
    pub fn new() -> FxRandomState;
}

impl BuildHasher for FxRandomState { ... }
```

This adds ways to create `FxHasher` with a custom or random seed.

Resolves #14
Resolves #15

Helps with #17, though different default seed may still be a good idea.

PureWhiteWu commented:
Hi, is there any plan for this pr to be merged?
I think this is quite important.

Nilstrieb commented on src/lib.rs:
```diff
@@ -46,6 +49,14 @@ pub type FxHashMap<K, V> = HashMap<K, V, BuildHasherDefault<FxHasher>>;
 
+/// Type alias for a hashmap using the `fx` hash algorithm with [`FxRandomState`].
+#[cfg(feature = "rand")]
+pub type FxHashMapRnd<K, V> = HashMap<K, V, FxRandomState>;
```
```suggestion
pub type FxHashMapRand<K, V> = HashMap<K, V, FxRandomState>;
```
I'd prefer writing out the `a`, or maybe dropping these aliases alltogether.

WaffleLapkin replied:
🆗

Nilstrieb reviewed (changes requested):
Can you add some tests for this?
One useful test would be to create two `FxRandomState` and ensure their seeds aren't the same.
Another possible test would be creating two `FxRandomState` on two different threads and ensuring their seeds aren't the same (this is not *guaranteed* to always pass, but should in practice always pass).
Another useful test would be for `FxHasher::with_seed` to use it with two different seeds (and the default) and ensure that hashing some integer with it produces different results in all cases.

WaffleLapkin commented:
@Nilstrieb are the tests I added match what you asked for?

I also added a `BuildHasher` impl you can provide a seed for, since there wasn't a way to do this without making your own type... Does it look reasonable to you?

Nilstrieb commented on src/lib.rs:
```diff
@@ -28,6 +28,13 @@
+extern crate rand;
+
+#[cfg(feature = "rand")]
+mod random_state;
```
Can we add a newline between the modules to make it more obvious that only the former is cfged?

Nilstrieb reviewed (approved):
r=me after style nit (also on the re-exports)

WaffleLapkin commented:
Uuuuhhhhh is the master branch not protected? D:

Nilstrieb commented:
I think it is, but I approved your PR, and GitHub does not dismiss approvals after pushing

WaffleLapkin commented:
@Nilstrieb I merged the PR before CI passed (or at least it looked like I did, maybe it happened at the same time?). also fyi github has an option to dismiss approvals after pushing

Nilstrieb commented:
interesting.. i wouldn't wanna dismiss approvals, that's just annoying, but if it's not gated on CI that would be a bit bad. not that it matters too much
